docs/proposals/daemon-controller.md

```diff
@@ -0,0 +1,151 @@
+#### Apiserver
+- Accept, parse, validate client commands
+- REST API calls will be handled in registry/daemoncontroller
+  - In particular, the api server will add the object to etcd
+  - DaemonManager listens for updates to etcd (using Framework.informer)
+- API object for Daemon Controller will be created in api/v1/types.go and api/v1/register.go
+#### DaemonManager
+- Creates new daemon controllers when requested. Launches the corresponding daemon pod on all nodes with labels matching the new daemon controller’s selector.
+- Listens for addition of new nodes to the cluster, by setting up a framework.NewInformer that watches for the creation of Node API objects. When a new node is added, the daemon manager will loop through each daemon controller. If the label of the node matches the selector of the daemon controller, then the daemon manager will create the corresponding daemon pod in the new node.
+- The daemon manager will create a pod on a node by sending a command to the API server, requesting for a pod to be bound to the node (the node will be specified via its hostname)
+#### Kubelet
+- Does not need to be modified, but health checking for the daemon pods and revive the pods if they are killed (we will set the pod restartPolicy to Always). We reject Daemon Controller objects with pod templates that don’t have restartPolicy set to Always.
+
+## Milestones
+
+1. Implement REST API skeleton code - users can add, update, delete (and perform other queries) on the Daemon Controller via the REST interface, but the API server won’t perform any action and will simply return a JSON object indicating that the query has been received.
+  1. Add code to API server that describes the type and specs of a daemon controller
+  2. Add code to validate that a daemon controller satisfies the specs
+2. Add Kubectl support. Kubectl queries should be handled and appropriately forwarded to the API server. Also “kubectl get dc <dc name>” needs to be implemented.
+3. User should be able to create daemon pods that are launched on every node in the cluster by calling the REST API or Kubectl.
+  1. Add code in the daemon manager that creates daemon controller objects, loops through every node in the cluster, and adds the specified pod to the node (via a call to API server)
+  2. New nodes added to the cluster will run daemon pods. This involves adding code to daemon manager that listens for node additions, and adds all daemons to the new node
+  3. Add support for labels (for launching daemon pods in exactly those nodes with a certain label).
+    1. Add an extra field to the spec of the Daemon controller that allows users to specify a node selector. Modify the code that validates the daemon controller spec accordingly.
```

Comment:
Oh yeah, that's what I'm doing. I updated the doc above, but I forgot to update this part of the doc.